syntax = "proto3";
import "google/protobuf/empty.proto";

package keyvaluestorage;


service StorageNodeStatus {
    // Function invoked to send the request
    rpc healthCheck (google.protobuf.Empty) returns (Ack) {}
}

message Ack {
  int32 num = 1;
}

service KeyValueStorage {
    // Function invoked to send the request
    rpc get (GetRequest) returns (GetResponse) {}
    rpc put (PutRequest) returns (PutResponse) {}
    rpc size (google.protobuf.Empty) returns (NodeSize) {}
}

service ControlInfo{
    rpc get_node_count(google.protobuf.Empty) returns (NodeCount){}
    rpc get_k_nodes(PutRequest) returns (StorageNodes){}
    rpc get_next_primary_node(GetRequest) returns(PutResponse) {}
}

message StorageNodes
{
    repeated string nodeid=1;
}

message NodeCount
{
    int32 numnodes=1;
}

message PutRequest
{
    string key=1;
    repeated string value=2;
}

message PutResponse
{
    string nodeid = 1;
}

message GetRequest
{
    string key=1;
}
message GetResponse
{
    repeated string value=1;
}

message NodeSize{
    int32 size=1;
}
